kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/KafkaProxyBackendHandler.java

```diff
@@ -5,99 +5,170 @@
  */
 package io.kroxylicious.proxy.internal;
 
+import java.util.Objects;
+import java.util.Optional;
+
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import io.netty.buffer.Unpooled;
 import io.netty.channel.Channel;
+import io.netty.channel.ChannelFutureListener;
 import io.netty.channel.ChannelHandlerContext;
 import io.netty.channel.ChannelInboundHandlerAdapter;
+import io.netty.handler.ssl.SslContext;
+import io.netty.handler.ssl.SslHandshakeCompletionEvent;
 
-import static java.util.Objects.requireNonNull;
+import io.kroxylicious.proxy.model.VirtualCluster;
+import io.kroxylicious.proxy.tag.VisibleForTesting;
 
 public class KafkaProxyBackendHandler extends ChannelInboundHandlerAdapter {
 
     private static final Logger LOGGER = LoggerFactory.getLogger(KafkaProxyBackendHandler.class);
 
-    private final KafkaProxyFrontendHandler frontendHandler;
-    private final ChannelHandlerContext inboundCtx;
-    private final KafkaProxyExceptionMapper exceptionMapper;
-    private ChannelHandlerContext blockedOutboundCtx;
-    private boolean unflushedWrites;
+    @VisibleForTesting
+    final ProxyChannelStateMachine proxyChannelStateMachine;
+    @VisibleForTesting
+    final SslContext sslContext;
+    ChannelHandlerContext serverCtx;
+    private boolean pendingServerFlushes;
 
-    public KafkaProxyBackendHandler(KafkaProxyFrontendHandler frontendHandler, ChannelHandlerContext inboundCtx, KafkaProxyExceptionMapper exceptionMapper) {
-        this.frontendHandler = frontendHandler;
-        this.inboundCtx = requireNonNull(inboundCtx);
-        this.exceptionMapper = exceptionMapper;
+    public KafkaProxyBackendHandler(
+                                    ProxyChannelStateMachine proxyChannelStateMachine,
+                                    VirtualCluster virtualCluster) {
+        this.proxyChannelStateMachine = Objects.requireNonNull(proxyChannelStateMachine);
+        Optional<SslContext> upstreamSslContext = virtualCluster.getUpstreamSslContext();
+        this.sslContext = upstreamSslContext.orElse(null);
     }
 
     @Override
     public void channelWritabilityChanged(final ChannelHandlerContext ctx) throws Exception {
         super.channelWritabilityChanged(ctx);
-        frontendHandler.outboundWritabilityChanged(ctx);
-    }
-
-    public void inboundChannelWritabilityChanged(ChannelHandlerContext inboundCtx) {
```

Comment:
Hmm the other source of confusion around this is netty invokes `channelWritabilityChanged` on the backend handler to tell us that the outbound/backend/server channel has changed state so the reads that need blocked/unblocked in response are actually the *clientReads*. Otherwise we could never recover from back pressure. 

I've very open to a discussion around making this easier to grok, so naming suggestions or diagrams etc.